Ep 10 zonal stats (#20)
esciencecenter-digital-skills/geospatial-python: _episodes/10-zonal-statistics.md

---
title: "Calculating Zonal Statistics on Rasters"
teaching: 40
exercises: 20
questions:

> - "How to rasterize a vector data-set?"
> - "How to calculate zonal statistics of a raster data-set?"

Author: ```suggestion
- "How to compute raster statistics on different zones delineated by a vector data?"
```

I changed it a bit because for me when you say over different parts of the vector data, it means the domain is defined by the vector data but it is actually the raster domain.

We convert the output of the `rasterio.features.rasterize` function, which generates a numpy array `np.ndarray`, to `xarray.DataArray` which will be used further:
~~~
field_cropped_raster_xarr = xr.DataArray(field_cropped_raster)
~~~
{: .language-python}

# Calculate zonal statistics 

In order to calculate zonal statistics we call the function, `xrspatial.zonal_stats`, which can calculate the minimum, maximum, mean, sum, median, variance, and standard deviation for each zone as defined by our vector data-set. The `xrspatial.zonal_stats` function takes as input `zones`, a 2D `xarray.DataArray`, that defines different zones, and `values`, a 2D `xarray.DataArray` providing input values for calculating statistics. We use the `squeeze()` function in order to reduce our raster data `ndvi` dimension to 2D by removing the singular `band` dimension:

~~~
ndvi_sq = ndvi.squeeze()
~~~
{: .language-python}

Then we call the `zonal stats` function with `cropfield_raster_xarr` as our classifier and the 2D raster with our values of interest `ndvi_sq` to obtain the NDVI statistics for each crop type:

~~~

> zonal_stats(cropfield_raster_xarr, ndvi_sq)

Author: ```suggestion
from xrspatial import zonal_stats
zonal_stats(cropfield_raster_xarr, ndvi_sq)
```

In this episode, we will explore how to calculate zonal statistics based on the types of crops in `cropped_field.shp` . To do this, we will first identify zones from the vector data, then rasterize these vector zones. Finally the zonal statistics for `ndvi` will be calculated over the rasterized zones.  

# Making vector and raster data compatible
Let's first read the crop fields data from our saved `cropped_field.shp` file and view the CRS information.

~~~
field = gpd.read_file('cropped_field.shp')
field.crs
~~~
{: .language-python}

~~~
<Derived Projected CRS: EPSG:28992>
Name: Amersfoort / RD New
Axis Info [cartesian]:
- X[east]: Easting (metre)
- Y[north]: Northing (metre)
Area of Use:
- name: Netherlands - onshore, including Waddenzee, Dutch Wadden Islands and 12-mile offshore coastal zone.
- bounds: (3.2, 50.75, 7.22, 53.7)
Coordinate Operation:
- name: RD New
- method: Oblique Stereographic
Datum: Amersfoort
- Ellipsoid: Bessel 1841
- Prime Meridian: Greenwich
~~~
{: .output}

> In order to use the vector data as a classifier for our raster, we need to convert the vector data to the appropriate CRS and crop to our raster domain. The CRS conversion can be done from the vector CRS (EPSG:28992) to our raster `ndvi` CRS (EPSG:32631) with:

Author: Indeed, it is better if it can be standalone - fixed by adding a read option for ndvi.

~~~
field_to_raster_crs = field.to_crs(ndvi.rio.crs)
~~~
{: .language-python}

> We can further crop the vector data with the same bounds as `ndvi` and view the data:

Author: I did a quick check and you are right. We do not need to crop the vector to the raster bounds to make it work. I have fixed this in the episode by removing the section on cropping.

 [<shapely.geometry.polygon.Polygon at 0x7ff88685ca00>, 331],
 ...]
~~~
{: .output}

This generates a list of the shapely geometries from the `geometry` column, and the unique field ID from the `gewascode` column in the `fields_cropped` geodataframe.

We can now rasterize our vector data using `rasterio.features.rasterize`:

~~~
from rasterio import features 
field_cropped_raster = rasterio.features.rasterize(geom, out_shape=ndvi_sq.shape, fill=0, transform=ndvi.rio.transform())
~~~
{: .language-python}

The argument `out_shape` specifies the shape of the output grid in pixel units, while `transform` represents the projection from pixel space to the projected coordinate space. We also need to specify the fill value for pixels that are not contained within a polygon in our shapefile, which we do with `fill = 0`. It's important to pick a fill value that is not the same as any value already defined in `gewascode` or else we won't distinguish between this zone and the background.

We convert the output of the `rasterio.features.rasterize` function, which generates a numpy array `np.ndarray`, to `xarray.DataArray` which will be used further:
~~~
import xarray as xr

> field_cropped_raster_xarr = xr.DataArray(field_cropped_raster)

Reviewer: This is actually `cropfield_raster_xarr` below - correct?

Author: Yes it is. Fixed now.

4	331	0.703056	0.905304	0.142226	10725.819336	0.102255	0.010456	15256.0
5	332	0.681699	0.849158	0.178113	321.080261	0.123633	0.015285	471.0
6	335	0.648063	0.865804	0.239661	313.662598	0.146582	0.021486	484.0
7	863	0.388575	0.510572	0.185987	1.165724	0.144245	0.020807	3.0
~~~
{: .output}

The `zonal_stats` function calculates the minimum, maximum, and sum for each zone along with statistical measures such as the mean, variance and standard deviation for each rasterized vector zone. In our raster data-set `zone = 0`, corresponding to non-crop areas, has the highest count followed by `zone = 265` which corresponds to 'Grasland, blijvend' or 'Grassland, permanent'. The highest mean NDVI is observed for `zone = 266` for 'Grasslands, temporary' with the lowest mean, aside from non-crop area, going to `zone = 863` representing 'Forest without replanting obligation'. Thus, the `zonal_stats` function can be used to analyse and understand different sections of our raster data. The definition of the zones can be derived from vector data or from classified raster data as presented in the challenge below: 

> ## Challenge: Calculate zonal statistics for zones defined by `ndvi_classified`
> 

> > Let's calculate NDVI zonal statistics for the different zones as classified by `ndvi_classified` in the previous episode .

Reviewer: Same as above, I think it would be better to not assume  `ndvi_classified` to be already loaded:
```suggestion
> Let's calculate NDVI zonal statistics for the different zones as classified and saved to `NDVI_classified.tif` in the previous episode.
```
(also need to add `ndvi_classified = rioxarray.open_rasterio("NDVI_classified.tif")` below)

Author: Resolved as before.
